File: lib/src/profiles/dtos/profile_dto.dart
class ProfileDto {
  final String username;
  final String? bio;
  final String? image;
  final bool following;

  ProfileDto(
      {required this.username, this.bio, this.image, required this.following});

  ProfileDto.fromJson(Map<String, dynamic> json)
      : username = json['profile']['username'],
        bio = json['profile']['bio'],
        image = json['profile']['image'],
        following = json['profile']['following'];

  Map<String, dynamic> toJson() => {
        'profile': {
          'username': username,
          'bio': bio,
          'image': image,
          'following': following
        }
      };
}

File: src/main/scala/com/softwaremill/realworld/profiles/ProfileDto.scala
package com.softwaremill.realworld.profiles

import com.softwaremill.realworld.common.NoneAsNullOptionEncoder.*
import zio.json.{DeriveJsonDecoder, DeriveJsonEncoder, JsonDecoder, JsonEncoder}

case class Profile(profile: ProfileData)

object Profile:
  given profileEncoder: JsonEncoder[Profile] = DeriveJsonEncoder.gen[Profile]
  given profileDecoder: JsonDecoder[Profile] = DeriveJsonDecoder.gen[Profile]

case class ProfileData(username: String, bio: Option[String], image: Option[String], following: Boolean)

object ProfileData:
  given profileDataEncoder: JsonEncoder[ProfileData] = DeriveJsonEncoder.gen[ProfileData]
  given profileDataDecoder: JsonDecoder[ProfileData] = DeriveJsonDecoder.gen[ProfileData]

case class Followers(userId: Int, followerId: Int)
